### DISCUSSION TITLE
Error Upgrading KIC to 2.9 and higher versions

### DISCUSSION THREAD

**Discussion by mayankmurari** [Kong/kubernetes-ingress-controller#6068] [Q&A]
We are trying to upgrade KIC to 2.9.x and eventually to 3.x. Current KIC version is 2.8.2 and Kong is running on version 3.6.1.

While upgrading Kong pod is not starting and we are getting below error in ingress-controller container.
`Error: unknown command "/kong-ingress-controller" for ""`

Can someone please help on this issue?

**Comment by pmalek** [comment]
Are you using a custom `command`? Currently the `ENTRYPOINT` in our image is `/manager` (ref: https://hub.docker.com/layers/kong/kubernetes-ingress-controller/3.1.5/images/sha256-4841452b4c4d1fab6ce529bbe6486c4766bec08e5ae4e9952174eb9923ddea14?context=explore) and `/kong-ingress-controller` does not exist in the image.

**Reply by mayankmurari** [reply]
@pmalek We are using helm charts for deployment. We are just pulling and deploying new image and everything else remains same.

**Reply by pmalek** [reply]
Can you provide the `values.yaml` that you use and which chart do you use for deployment?

**Reply by mayankmurari** [reply]
Below is the section from values files. Only change is in image tag from 2.8.2 to 2.9.0.
ingressController:
    enabled: "true"
    imageRepository: repo_host/kong/kubernetes-ingress-controller
    imageTag: "2.9.0"

**Reply by pmalek** [reply]
Both [2.8.2](https://hub.docker.com/layers/kong/kubernetes-ingress-controller/2.8.2/images/sha256-9e170e1cc666949dc2aa29f1a9af1f5eab1d5105338fee5f0232ea33ab3ece13?context=explore) and [2.9](https://hub.docker.com/layers/kong/kubernetes-ingress-controller/2.9.0/images/sha256-3e6a34c36e6f36e58d5446d704cee23e1505bb45f15d2c68fb5be7e0fee2bb11?context=explore) have `/manager` as entrypoint. I'd assume there's something that changes the entrypoint (`command` in k8s) in your env.

Our images do not mention or use `/kong-ingress-controller` as entrypoint.

**Reply by mayankmurari** [reply]
Hi @pmalek 
Just to add some more context we are using v0.9.0 of operator and can see below error in operator logs.

`{"level":"info","ts":1716293056.1918042,"logger":"helm.controller","msg":"Reconciled release","namespace":"hawkeye-ns","name":"hawkeye","apiVersion":"charts.konghq.com/v1alpha1","kind":"Kong","release":"hawkeye"}
{"level":"error","ts":1716293056.5803494,"logger":"controller-runtime.controller","msg":"Reconciler error","controller":"kong-controller","request":"hawkeye-ns/hawkeye","error":"Operation cannot be fulfilled on kongs.charts.konghq.com \"hawkeye\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\tpkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:258\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\tpkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\tpkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\tpkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:88"}`

Also we can see "/kong-ingress-controller" args passed here.
https://github.com/Kong/kong-operator/blob/main/helm-charts/kong/templates/_helpers.tpl#L571

Do you know 0.9.0 version of kong-operator will have compatibility till which version of KIC?
https://github.com/Kong/kong-operator

**Reply by pmalek** [reply]
As you can see in the readme in https://github.com/Kong/kong-operator, kong-operator has been deprecated a while ago in favor of recently released https://github.com/Kong/gateway-operator/.

**Reply by mayankmurari** [reply]
@pmalek Yes we are aware of new Go Based operator and planning to migrate to it, but trying to understand if this error is because of compatibility with existing operator. It's not mentioned till which version of KIC existing kong-operator will be compatible which we can consider as hard stop.

**Reply by pmalek** [reply]
Last release of kong-operator (v0.9.0 which is what you've mentioned you're using) has been published in Oct 2021, which is around the time [KIC v2.0](https://github.com/Kong/kubernetes-ingress-controller/blob/main/CHANGELOG.md/#200) has been released.

We do not publish the compatibility matrix between kong-operator and latest KIC releases. 

KGO (Kong gateway operator) does have the compatibility matrix published on the product page: https://docs.konghq.com/gateway-operator/latest/reference/version-compatibility/#kubernetes.

If you'd like to get help on migration to KGO please create a discussion in https://github.com/Kong/gateway-operator/.

**Reply by mayankmurari** [reply]
@pmalek Thanks for your inputs. We will plan migrating to Kong Gateway Operator.
